code/WorkInProgress/blueprints.dm

```diff
@@ -384,127 +385,338 @@
+			return
+		// ckeyEx to sanitize filename: no spaces/special chars, only '_', '-', and '@' allowed. 54 char limit in tgui_input
+		var/input = ckeyEx(tgui_input_text(user, "You are copying '[src.room_name]' to your own collection. \
+			Choose a file name for it. Use only alphanumeric characters, and - and _.", "Copy Homework", ckeyEx(src.room_name), 54))
+		var/timeout = 0
+		while (input && fexists("data/blueprints/[user.client.ckey]/[input].dat"))
+			if (!user?.client?.ckey || timeout > 5)
+				boutput(user, "<span class='alert'>Copy operation timed out. Please try again.</span>")
+				return
+			input = ckeyEx(tgui_input_text(user, "A blueprint named '[input]' already exists. Please input another, or cancel.",
+				"Copy Homework", input, 54)) // handy dandy prompt autofilled with the last used input
+			timeout++
+		if (!input) return
+		fcopy(src.blueprint_path, "data/blueprints/[user.client.ckey]/[input].dat")
+		boutput(user, "<span class='notice'>Copied this blueprint! Its filename is: '[input]'.</span>")
+
+	afterattack(atom/target, mob/user)
+		if (!istype(target, /obj/machinery/abcu))
+			. = ..()
+			return
+		if (!src.blueprint_path || !fexists(src.blueprint_path))
+			boutput(user, "<span class='alert'>This item is broken, please tell a coder if it keeps breaking!</span>")
+			return
+		var/obj/machinery/abcu/abcu = target
+		abcu.get_blueprint(user, src.blueprint_path)
+
+// whitelists/blacklists applied during both saving and loading, so it's functionally retroactive
+#define WHITELIST_OBJECTS list( \
+	/obj/stool, \
+	/obj/grille, \
+	/obj/window, \
+	/obj/machinery/door, \
+	/obj/cable, \
+	/obj/table, \
+	/obj/rack, \
+	/obj/structure, \
+	/obj/disposalpipe, \
+	/obj/machinery/light, \
+	/obj/machinery/door_control, \
+	/obj/machinery/light_switch, \
+	/obj/machinery/camera, \
+	/obj/item/device/radio/intercom, \
+	/obj/machinery/firealarm, \
+	/obj/machinery/power/apc, \
+	/obj/machinery/alarm, \
+	/obj/machinery/disposal, \
+	/obj/machinery/gibber, \
+	/obj/machinery/floorflusher, \
+	/obj/machinery/activation_button/driver_button, \
+	/obj/machinery/door_control, \
+	/obj/machinery/disposal, \
+	/obj/submachine/chef_oven, \
+	/obj/submachine/chef_sink, \
+	/obj/machinery/launcher_loader, \
+	/obj/machinery/optable, \
+	/obj/machinery/mass_driver, \
+	/obj/machinery/sleeper, \
+	/obj/machinery/sleep_console, \
+	/obj/submachine/slot_machine, \
+	/obj/machinery/deep_fryer, \
+	/obj/submachine/ATM, \
+	/obj/submachine/ice_cream_dispenser, \
+	/obj/machinery/portable_atmospherics, \
+	/obj/machinery/ai_status_display, \
+	/obj/securearea, \
+	/obj/submachine/mixer, \
+	/obj/submachine/foodprocessor, \
+)
+// blacklist overrules whitelist
+#define BLACKLIST_OBJECTS list( \
+	/obj/disposalpipe/loafer, \
+	/obj/submachine/slot_machine/item, \
+	/obj/machinery/portable_atmospherics/canister, \
+)
+
+#define WHITELIST_TURFS list(/turf/simulated)
+
+/datum/abcu_blueprint
```

Comment:
can you move this to a different file than the abcu and all of this to the `modules/blueprints` folder or something? it's way due for organization

Reply:
Im glad you asked. How much harder will that make the diff to read?

Reply:
it shouldn't, really